src/ucp/rndv/rndv_rkey_ptr.c

```diff
@@ -183,3 +190,266 @@ ucp_proto_t ucp_rndv_rkey_ptr_proto = {
+    ucp_md_map_t mdesc_md_map;
+    size_t frag_size;
+
+    ucp_md_index_t md_index;
+    const uct_md_attr_t *md_attr;
+    int found_md_map = 0;
+
+    if (!context->config.ext.rndv_shm_ppln_enable) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if ((init_params->select_param->dt_class != UCP_DATATYPE_CONTIG) ||
+        !ucp_proto_rndv_op_check(init_params, UCP_OP_ID_RNDV_SEND, 1) ||
+        init_params->rkey_config_key == NULL) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    status = ucp_proto_rndv_mtype_init(init_params, &mdesc_md_map, &frag_size);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucs_for_each_bit(md_index, init_params->rkey_config_key->md_map) {
+        md_attr = &context->tl_mds[md_index].attr;
+        if ((md_attr->cap.flags & UCT_MD_FLAG_RKEY_PTR) &&
+            /* Do not use xpmem, because cuda_copy registration will fail and
+             * performance will not be optimal. */
+            !(md_attr->cap.flags & UCT_MD_FLAG_REG) &&
+            (md_attr->cap.access_mem_types &
+             UCS_BIT(init_params->rkey_config_key->mem_type))) {
+            found_md_map = UCS_BIT(md_index);
+            break;
+        }
+    }
+    if (found_md_map == 0) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return ucp_proto_rndv_rkey_ptr_mtype_init_params(init_params, mdesc_md_map,
+                                                     frag_size);
+}
+
+
+static void ucp_proto_rndv_rkey_ptr_mtype_completion(uct_completion_t *uct_comp)
+{
+    ucp_request_t *req = ucs_container_of(uct_comp, ucp_request_t,
+                                          send.state.uct_comp);
+
+    ucp_trace_req(req, "ucp_proto_rndv_rkey_ptr_mtype_completion");
+    ucp_proto_rndv_rkey_destroy(req);
+    ucp_proto_request_zcopy_complete(req, req->send.state.uct_comp.status);
+}
+
+static void
+ucp_proto_rndv_rkey_ptr_mtype_copy_completion(uct_completion_t *uct_comp)
+{
+    ucp_request_t *req = ucs_container_of(uct_comp, ucp_request_t,
+                                          send.state.uct_comp);
+
+    ucp_trace_req(req, "ucp_proto_rndv_rkey_ptr_mtype_copy_completion");
+
+    ucp_proto_completion_init(&req->send.state.uct_comp,
+                              ucp_proto_rndv_rkey_ptr_mtype_completion);
+    ucp_proto_request_set_stage(req, UCP_PROTO_RNDV_RKEY_PTR_STAGE_ATP);
+    ucp_request_send(req);
+}
+
+
+static ucs_status_t
+ucp_proto_rndv_rkey_ptr_mtype_copy_progress(uct_pending_req_t *uct_req)
+{
+    ucp_request_t *req                = ucs_container_of(uct_req, ucp_request_t,
+                                                         send.uct);
+    ucp_context_h context             = req->send.ep->worker->context;
+    ucp_md_index_t rkey_ptr_md_index  = UCP_NULL_RESOURCE;
+    uint64_t remote_address           = req->send.rndv.remote_address;
+    ucs_memory_type_t local_mem_type  = req->send.state.dt_iter.mem_info.type;
+    const ucp_proto_rndv_put_priv_t *rpriv = req->send.proto_config->priv;
```

Comment:
we do not really do PUT here, probably need to use another proto-specific structure.